File: sql-client-examples/src/main/ruby/io/vertx/example/sqlclient/query_params/sql_client_example.rb
require 'vertx-pg-client/pg_pool'
require 'vertx-sql-client/tuple'

pool = VertxPgClient::PgPool.pool($vertx, {
  'port' => 5432,
  'host' => "the-host",
  'database' => "the-db",
  'user' => "user",
  'password' => "secret"
}, {
  'maxSize' => 4
})

# Uncomment for MySQL
#    Pool pool = MySQLPool.pool(vertx, new MySQLConnectOptions()
#      .setPort(5432)
#      .setHost("the-host")
#      .setDatabase("the-db")
#      .setUser("user")
#      .setPassword("secret"), new PoolOptions().setMaxSize(4));

pool.get_connection() { |res1_err,res1|
  if (res1_err != nil)
    STDERR.puts res1_err.get_message()
    return
  end
  connection = res1

  # create a test table
  connection.query("create table test(id int primary key, name varchar(255))") { |res2_err,res2|
    if (res2_err != nil)
      connection.close()
      STDERR.puts "Cannot create the table"
      res2_err.print_stack_trace()
      return
    end

    # insert some test data
    connection.query("insert into test values (1, 'Hello'), (2, 'World')") { |res3_err,res3|

      # query some data with arguments
      connection.prepared_query("select * from test where id = ?", VertxSqlClient::Tuple.of(2)) { |rs_err,rs|
        if (rs_err != nil)
          STDERR.puts "Cannot retrieve the data from the database"
          rs_err.print_stack_trace()
          return
        end

        rs.each do |line|
          puts "#{line}"
        end

        # and close the connection
        connection.close()
      }
    }
  }
}

File: sql-client-examples/src/main/java/io/vertx/example/sqlclient/query_params/SqlClientExample.java
package io.vertx.example.sqlclient.query_params;

import io.vertx.core.AbstractVerticle;
import io.vertx.core.Vertx;
import io.vertx.mysqlclient.MySQLConnectOptions;
import io.vertx.pgclient.PgConnectOptions;
import io.vertx.sqlclient.Pool;
import io.vertx.sqlclient.PoolOptions;
import io.vertx.sqlclient.Row;
import io.vertx.sqlclient.SqlConnectOptions;
import io.vertx.sqlclient.Tuple;
import org.testcontainers.containers.MySQLContainer;
import org.testcontainers.containers.PostgreSQLContainer;

/*
 * @author <a href="mailto:pmlopes@gmail.com">Paulo Lopes</a>
 */
public class SqlClientExample extends AbstractVerticle {

  // Convenience method so you can run it in your IDE
  public static void main(String[] args) {
    PostgreSQLContainer<?> postgres = new PostgreSQLContainer<>();
    postgres.start();
    PgConnectOptions options = new PgConnectOptions()
      .setPort(postgres.getMappedPort(5432))
      .setHost(postgres.getContainerIpAddress())
      .setDatabase(postgres.getDatabaseName())
      .setUser(postgres.getUsername())
      .setPassword(postgres.getPassword());
    // Uncomment for MySQL
//    MySQLContainer<?> mysql = new MySQLContainer<>();
//    mysql.start();
//    MySQLConnectOptions options = new MySQLConnectOptions()
//      .setPort(mysql.getMappedPort(3306))
//      .setHost(mysql.getContainerIpAddress())
//      .setDatabase(mysql.getDatabaseName())
//      .setUser(mysql.getUsername())
//      .setPassword(mysql.getPassword());
    Vertx vertx = Vertx.vertx();
    vertx.deployVerticle(new SqlClientExample(options));
  }

  private final SqlConnectOptions options;

  public SqlClientExample(SqlConnectOptions options) {
    this.options = options;
  }

  @Override
  public void start() {

    Pool pool = Pool.pool(vertx, options, new PoolOptions().setMaxSize(4));

    // create a test table
    pool.query("create table test(id int primary key, name varchar(255))")
      .execute()
      .compose(r ->
        // insert some test data
        pool
          .query("insert into test values (1, 'Hello'), (2, 'World')")
          .execute()
      ).compose(r ->
      // query some data with arguments
      pool
        .preparedQuery("select * from test where id = $1")
        .execute(Tuple.of(2))
    ).onSuccess(rows -> {
      for (Row row : rows) {
        System.out.println("row = " + row.toJson());
      }
    }).onFailure(Throwable::printStackTrace);
  }
}
